src/finchlite/autoschedule/einsum.py

```diff
@@ -0,0 +1,151 @@
+from collections.abc import Callable
+from typing import Any
+
+import finchlite.finch_einsum as ein
+from finchlite.algebra import init_value, is_commutative, overwrite
+import finchlite.finch_logic as lgc
+from finchlite.finch_logic import LogicNode
+
+class EinsumLowerer:
+    def __call__(self, prgm: lgc.Plan) -> tuple[ein.Plan, dict[str, Any]]:
+        bindings: dict[str, Any] = {}
+        definitions: dict[str, ein.Einsum] = {}
+        return self.compile_plan(prgm, bindings, definitions), bindings
+
+    def compile_plan(
+        self, plan: lgc.Plan, bindings: dict[str, Any], definitions: dict[str, ein.Einsum]
+    ) -> ein.Plan:
+        bodies: list[ein.EinsumNode] = []
+
+        for body in plan.bodies:
+            match body:
+                case lgc.Plan(_):
+                    bodies.append(self.compile_plan(body, bindings, definitions))
+                case lgc.Query(lgc.Alias(name), lgc.Table(lgc.Literal(val), _)):
+                    bindings[name] = val
+                case lgc.Query(
+                    lgc.Alias(name), lgc.Aggregate(lgc.Literal(operation), lgc.Literal(init), arg, idxs)
+                ) | lgc.Query(
+                    lgc.Alias(name),
+                    lgc.Aggregate(lgc.Literal(operation), lgc.Literal(init), lgc.Reorder(arg, _), idxs),
+                ):
+                    einidxs = tuple(ein.Index(field.name) for field in idxs)
+                    if init != init_value(operation, type(init)):
+                        bodies.append(
```

Comment:
instead of appending to a list of bodies, return either an einsum or a plan of einsums.